### PR TITLE
Issue #9142: Migrate to Truth in AbstractModuleTestSupport

### PR DESCRIPTION
**Pull request by shashwatj07**
Issue #9142

No Violations Expected:
```
Files with expected violations and actual violations differ.
value of: keySet()
expected to be empty
but was : [C:\Users\Shashwat\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\annotation\missingoverride\InputMissingOverrideBadOverrideFromObject.java]
```
Violations don't match:
```
Violations for C:\Users\Shashwat\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\annotation\missingoverride\InputMissingOverrideBadOverrideFromObject.java differ.
unexpected (1): 53:5: Must include @java.lang.Override annotation when {@inheritDoc} Javadoc tag exists.
---
expected      : [15:5: Must include @java.lang.Override annotation when {@inheritDoc} Javadoc tag exists., 35:9: Must include @java.lang.Override annotation when {@inheritDoc} Javadoc tag exists., 44:5: Must include @java.lang.Override annotation when {@inheritDoc} Javadoc tag exists.]
but was       : [15:5: Must include @java.lang.Override annotation when {@inheritDoc} Javadoc tag exists., 35:9: Must include @java.lang.Override annotation when {@inheritDoc} Javadoc tag exists., 44:5: Must include @java.lang.Override annotation when {@inheritDoc} Javadoc tag exists., 53:5: Must include @java.lang.Override annotation when {@inheritDoc} Javadoc tag exists.]
```

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:369]: I want a more fluent approach to this. Any suggestions?

Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:369]: Maybe 
```java
assertWithMessage(message.toString())
    .that(missingViolations)
    .isEmpty();
assertWithMessage(message.toString())
    .that(unexpectedViolations)
    .isEmpty();
// so on
```
?

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: Done this. However, I was looking for something more abstract.

**Comment by shashwatj07** [comment]
CI failed. I'm unable to see wercker logs, ```mvn clean verify``` is passing.

**Comment by AkMo3** [comment]
@shashwatj07 Please rebase, wrecker is fixed in #10025

**Review by pbludov — APPROVED** [review]

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (strkkk) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:373]: `message.toString()` please extract variable instead of repeating

**Comment by strkkk** [comment]
@shashwatj07 @pbludov may be it is better to make some util method like
`assertEmpty(Collection col, String message)` and reuse it instead of repeating boilerplate like
```
        assertWithMessage(message.toString()) 
             .that(unexpectedViolations) 
             .isEmpty();
```

I guess if this is going to be refactored in whole project it will be clearer.

**Comment by shashwatj07** [comment]
> @shashwatj07 @pbludov may be it is better to make some util method like
> `assertEmpty(Collection col, String message)` and reuse it instead of repeating boilerplate like
> 
> ```
>         assertWithMessage(message.toString()) 
>              .that(unexpectedViolations) 
>              .isEmpty();
> ```
> 
> I guess if this is going to be refactored in whole project it will be clearer.

Will that not hamper the fluency which Truth provides?

**Comment by shashwatj07** [comment]
Option 2:
``` assertAreEmpty(Collections ... cols, String message) ```

This may be more versatile for our use cases.

**Comment by strkkk** [comment]
> Will that not hamper the fluency which Truth provides?

what do you mean? it will just reduce similar code snippets

**Comment by shashwatj07** [comment]
> what do you mean? it will just reduce similar code snippets

The entire idea of Truth is to make assertions more fluent and English-like. Doing that is just abstracting it away.

**Comment by strkkk** [comment]
> The entire idea of Truth is to make assertions more fluent and English-like. Doing that is just abstracting it away.

Yeah, but I hate repeating same code over and over.
We can discuss it somewhere else, may be other maintainers have some ideas.

**Comment by shashwatj07** [comment]
> We can discuss it somewhere else, may be other maintainers have some ideas.

so what do I do for this one right now?

**Comment by strkkk** [comment]
> so what do I do for this one right now?
This is not a blocker for this, just a thought to discuss.
Please take a look at https://github.com/checkstyle/checkstyle/pull/10026#discussion_r637804550

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:364]: Can we do away with all the Maps and simply compare ```actualViolations``` to be equal to ```realExpectedViolations``` instead?

```
        assertWithMessage("Violations should be identical.")
                .that(actualViolations)
                .containsExactlyEntriesIn(realExpectedViolations)
```

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:364]: It is actually ```containsExactlyEntriesIn(``` for maps. I tried but it's not working as expected.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:376]: Also, this might end up logging the same thing thrice.

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: fixed

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: I think it makes it clear now.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: >  tried but it's not working as expected.

 I will use maps in my example so that we can discuss the output (though most of the time, `verify()` receives a Map with a single entry, so it would look more like above):

```bash

Violations should be identical.
keys with wrong values
for key:
    MyOtherFile.java
expected value:
    11:17: '42' is a magic number.
    13:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
but got value:
    11:17: '42' is a magic number.
    12:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
for key:
    MyFile.java
expected value:
    6: First sentence of Javadoc is missing an ending period.
    9: First sentence of Javadoc is missing an ending period.
    36: First sentence of Javadoc is missing an ending period.
    
but got value:
    6: First sentence of Javadoc is missing an ending period.
    7: First sentence of Javadoc is missing an ending period.
    17: First sentence of Javadoc is missing an ending period.
    27: First sentence of Javadoc is missing an ending period.
    36: First sentence of Javadoc is missing an ending period.
    
---
expected:
    MyOtherFile.java:
    11:17: '42' is a magic number.
    13:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
    MyFile.java:
    6: First sentence of Javadoc is missing an ending period.
    9: First sentence of Javadoc is missing an ending period.
    36: First sentence of Javadoc is missing an ending period.
    
    
but was:
    MyOtherFile.java:
    11:17: '42' is a magic number.
    12:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
    MyFile.java:
    6: First sentence of Javadoc is missing an ending period.
    7: First sentence of Javadoc is missing an ending period.
    17: First sentence of Javadoc is missing an ending period.
    27: First sentence of Javadoc is missing an ending period.
    36: First sentence of Javadoc is missing an ending period.

```

Please share what is unexpected here; maybe we could improve this idea.

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: I'll look at it again. I found that it was passing even when I flipped the order of violations in an input file.

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: The ordering was creating a fuss. Fixed and done now.

Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: Can you show a failure in CI, and share a link here? It would be good to see failed assertion format.

Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: Hm. I now notice that `containsExactlyEntriesIn()` does not give the diff that `containsExactlyElementsIn()`. Perhaps it is better to do:

```java

        expectedViolationsMap.forEach((k,v) -> {
            assertWithMessage("Violations should be identical.")
                .that(v)
                .containsExactlyElementsIn(actualViolationsMap.get(k))
                .inOrder();
        });

```

Yielding:

```bash

Violations should be identical.
missing (1):
    12:17: '42' is a magic number.
unexpected (1):
    18:17: '42' is a magic number.
---
expected:
    11:17: '42' is a magic number.
    12:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
but was:
    11:17: '42' is a magic number.
    14:17: '42' is a magic number.
    18:17: '42' is a magic number.

```

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: But what if the key k does not exist in ```actualViolationsMap```? ```actualViolationsMap.get(k)``` will return NULL.

Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: This is why I added 
```java

        assertWithMessage("Files in expected and actual differ.")
            .that(actualViolationsMap.keySet())
            .isEqualTo(expectedViolationsMap.keySet());

```
to my example :)

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: Got it

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: done

Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: > The entire idea of Truth is to make assertions more fluent and English-like.

I agree with this 100%. I put comments in my example for discussion's sake; you should remove them. I think that our intention is very clear now.

Instead of `key` and `value`, what do you think about `filename` and `violationList`?

Instead of  `assertWithMessage("Violations should be identical.")`, what do you think about including the filename somehow? In the case of multiple files, it might not be easy to determine which file has differing violations. Maybe `Violations for 'MyFile.java' differ:`?

Comment (AkMo3) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: @shashwatj07 @nmancus1 Intellij offers to compare the strings and view the differences as shown in [viewing-and-exploring-test-results.html](https://www.jetbrains.com/help/idea/viewing-and-exploring-test-results.html#assert-equals) . It comes really handy when there is a number of huge violation messages. This feature might be in other IDEs as well. Right now the "difference viewer" option is not showing up. Can we integrate it into our code as well?

Something like 

```
// check violations
realExpectedViolations.forEach((filename, violations) -> {
    assertWithMessage("Violations in \"%s\" should be identical.", filename)
        .that(actualViolations.get(filename).toArray())
        .isEqualTo(violations.toArray());
});
```

The output remains same just I added the file name to message:
```
Violations in "H:\GitHub\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\imports\unusedimports\InputUnusedImportsCheckClearState.java" should be identical.
expected        : [3:8: Unused import - java.util.Arrays., 4:12: Unused import - java.util.List., 5:9: Unused import - java.util.Set.]
but was         : [3:8: Unused import - java.util.Arrays., 4:8: Unused import - java.util.List., 5:8: Unused import - java.util.Set.]
differs at index: [1]
Expected :[3:8: Unused import - java.util.Arrays., 4:12: Unused import - java.util.List., 5:9: Unused import - java.util.Set.]
Actual   :[3:8: Unused import - java.util.Arrays., 4:8: Unused import - java.util.List., 5:8: Unused import - java.util.Set.]
```

I am getting the "difference viewer" option with this change, also it should not change the test as well according to my understanding. Please take a look and share your views.

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: @nmancus1 sure, done!

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: @AkMo3 From the link it seems that comparing differences is only available for ```string assertEquals failures```. Is it so?

Comment (AkMo3) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: @shashwatj07 No. `assertEquals` is just an example. As you can see below, I am getting _**click to see difference**_ option
image

Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: > Please take a look and share your views.

@AkMo3 I think the current implementation addresses the comments at https://github.com/checkstyle/checkstyle/issues/9142#issuecomment-848891011 and takes advantage of Truth's awesome assertion output (thus making the 'migration' complete). I do not hold want to hold up this PR any longer :)

@pbludov implementation has changed significantly, please take another look at this PR. What do you think about overriding `toString()` to make output like https://github.com/checkstyle/checkstyle/pull/10026#discussion_r641577310 (multiple lines) ?

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:308]: @shashwatj07 one more thought:

Couldn't the entire contents of this method be replaced by:

```java

        Map<String, List<String>> expectedViolations = new HashMap<>();
        expectedViolations.put(messageFileName, Arrays.asList(expected));
        verify(checker, processedFiles, expectedViolations);

```
?

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:308]: You're right. I've faced the same nuisance often times too. I would have made the changes myself if it weren't for this line here. What do you think?
https://github.com/checkstyle/checkstyle/blob/db5311e79c2f08fe4e2c46565370bf2c792ff684/src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java#L273-L274

Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:308]: It appears that this was added in https://github.com/checkstyle/checkstyle/pull/5236, but I do not see any discussion surrounding it. Considering that in both cases (for either `verify()` method that we are using), we are providing MORE information than was provided before, in a better format, I can't imagine why this would not be a good way to reuse existing code and improve debugging experience for maintainers.

FWIW, I think that it would be a good idea in this PR, and any future PR's for "Migrate to Truth..." where assertions are not simple, to temporarily modify a unit test so that it fails, then show a "before" and "after" in the PR description so that reviewers can compare between the output that Truth provides, and the old Hamcrest, JUnit, etc. assertions.

Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: done, sorry for the delay

**Comment by romani** [comment]
@shashwatj07 , please rebase to latest code to resolve CI problem.

@nmancus1 , please finish review

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:275]: I think that we would be better served by slightly improving the `@param` tags and keeping this description to one or two sentences.  Perhaps something like: `Performs verification of expected violations on given files.` or something like that. `This implementation uses overloaded...` seems redundant or unnecessary; I am not sure that anyone needs to know that to use this method.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:316]: Perhaps we could improve this a bit and do something like `Files with expected violations and actual violations differ.`

**Review by nmancus1 — CHANGES_REQUESTED** [review]
A couple of minor things after using this for a bit ( I cherry picked this PR to help debug :) ):

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: Resolved.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: This can be resolved.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:275]: The javadoc of other verify methods has the same structure (and redundancy). Should all be modified?

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:275]: It's up to you. If you change the others, I would make it a separate commit (`doc`).

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: Resolved.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:275]: Sure, I'll do it in the next PR.
@romani please verify and open issue.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java:275]: Please remove `This implementation uses  ...`
It is not required, not sure how all other methods get such description.
Javadoc should not explain logic of code if it is easy to read logic

**Review by romani — CHANGES_REQUESTED** [review]
Please remove suppression on this file from config, after merge of this PR all attempt to introduce junit should be violated.

As soon as import is removed, we should remove file from suppression.

**Comment by romani** [comment]
@strkkk , please create issue in this with your proposal and assign to this gsoc project, we will review it at the end of project.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (shashwatj07) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: done

**Comment by shashwatj07** [comment]
> Please remove suppression on this file from config

@romani it is not suppressed because it uses ```org.junit.Assert``` instead of ```org.junit.jupiter.api.Assertions``` which we have restricted using import control. Shall I open issue to restrict the former too?

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java]: Resolved.

**Review by nmancus1 — APPROVED** [review]
@shashwatj07 Please see comments at https://github.com/checkstyle/checkstyle/pull/10026#discussion_r642067373 . I would recommend that you share the new output of failed unit tests in PR description for the next reviewers to see. I would probably do a few cases, such as (1) no violations expected, but there are some, (2) violations are expected but do not match, (3) files with expected violations do not match.

**Comment by shashwatj07** [comment]
> I would recommend that you share the new output of failed unit tests in PR description for the next reviewers to see.

done

**Comment by strkkk** [comment]
@shashwatj07 there is a merge conflict, please resolve

**Review by strkkk — APPROVED** [review]

**Pull request merged**
